src/internal/connector/exchange/service_iterators.go

```diff
@@ -552,9 +499,109 @@ func FetchMessageIDsFromDirectory(
 		Messages().
 		Delta()
 
-	if err := getIDs(builder); err != nil {
+	added, removed, deltaURL, err := getItemsAddedAndRemovedFromContainer(
+		ctx,
+		&messagePager{gs, builder, options},
+		errUpdater,
+	)
+	if err != nil {
 		return nil, nil, deltaUpdate{}, err
 	}
 
-	return ids, removedIDs, deltaUpdate{deltaURL, resetDelta}, errs.ErrorOrNil()
+	return added, removed, deltaUpdate{deltaURL, resetDelta}, errs.ErrorOrNil()
+}
+
+// ---------------------------------------------------------------------------
+// generic handler for paging item ids in a container
+// ---------------------------------------------------------------------------
+
+type itemPager interface {
+	getPage(context.Context) (pageLinker, error)
+	setNext(string)
+}
+
+type pageLinker interface {
+	GetOdataDeltaLink() *string
+	GetOdataNextLink() *string
+}
+
+type getIDAndAddtler interface {
+	GetId() *string
+	GetAdditionalData() map[string]any
+}
+
+// generic controller for retrieving all item ids in a container.
+func getItemsAddedAndRemovedFromContainer(
+	ctx context.Context,
+	pager itemPager,
+	errUpdater func(error),
+) ([]string, []string, string, error) {
+	var (
+		addedIDs   = []string{}
+		removedIDs = []string{}
+		deltaURL   string
+	)
+
+	for {
+		// get the next page of data, check for standard errors
+		resp, err := pager.getPage(ctx)
+		if err != nil {
+			if err := graph.IsErrDeletedInFlight(err); err != nil {
+				return nil, nil, deltaURL, err
+			}
+
+			if err := graph.IsErrInvalidDelta(err); err != nil {
+				return nil, nil, deltaURL, err
+			}
+
+			return nil, nil, deltaURL, errors.Wrap(err, support.ConnectorStackErrorTrace(err))
+		}
+
+		// each category type responds with a different interface, but all
+		// of them comply with GetValue, which is where we'll get our item data.
+		gv, ok := resp.(interface{ GetValue() []getIDAndAddtler })
```

Comment:
this should be added to the `pageLinker` interface. Yes, it will require a bit more glue code to convert the slice return values, but at least at that point it's clear what the interface actually requires as opposed to having a failure later on on a seemingly unrelated line of code even though the `pageLinker` interface was properly implemented according to the definition

Reply:
Hah, it's funny to have you mention that, as I went through a process of removing that extra glue for slimness.  Good points, though.